pkg/render/common/components/components.go

```diff
@@ -0,0 +1,176 @@
+			nodeSelector = obj.Spec.Template.Spec.NodeSelector
+
+			if len(obj.Spec.Template.Spec.InitContainers) > 0 {
+				for _, v := range obj.Spec.Template.Spec.InitContainers {
+					c := container{Name: v.Name, Resources: v.Resources}
+					initContainers = append(initContainers, c)
+				}
+			}
+			if len(obj.Spec.Template.Spec.Containers) > 0 {
+				for _, v := range obj.Spec.Template.Spec.Containers {
+					c := container{Name: v.Name, Resources: v.Resources}
+					containers = append(containers, c)
+				}
+			}
+		}
+	}
+
+	if metadata != nil {
+		if len(metadata.Labels) > 0 {
+			if ds.Labels == nil {
+				ds.SetLabels(make(map[string]string))
+			}
+			mergeMaps(ds.Labels, metadata.Labels)
+		}
+		if len(metadata.Annotations) > 0 {
+			if ds.GetAnnotations() == nil {
+				ds.SetAnnotations(make(map[string]string))
+			}
+			mergeMaps(ds.GetAnnotations(), metadata.Annotations)
+		}
+	}
+
+	if minReadySeconds != nil {
+		ds.Spec.MinReadySeconds = *minReadySeconds
+	}
+
+	if podTemplateMetadata != nil {
+		if len(podTemplateMetadata.Labels) > 0 {
+			if ds.Spec.Template.GetLabels() == nil {
+				ds.Spec.Template.SetLabels(make(map[string]string))
+			}
+			mergeMaps(ds.Spec.Template.GetLabels(), podTemplateMetadata.Labels)
+		}
+		if len(podTemplateMetadata.Annotations) > 0 {
+			if ds.Spec.Template.GetAnnotations() == nil {
+				ds.Spec.Template.SetAnnotations(make(map[string]string))
+			}
+			mergeMaps(ds.Spec.Template.GetAnnotations(), podTemplateMetadata.Annotations)
+		}
+	}
+
+	// Merge all other fields in the overrides spec if they exist.
+	if initContainers != nil {
+		mergeContainers(ds.Spec.Template.Spec.InitContainers, initContainers)
+	}
+	if containers != nil {
+		mergeContainers(ds.Spec.Template.Spec.Containers, containers)
+	}
+	if affinity != nil {
+		ds.Spec.Template.Spec.Affinity = affinity
+	}
+	if nodeSelector != nil {
+		ds.Spec.Template.Spec.NodeSelector = nodeSelector
+	}
+	if tolerations != nil {
+		ds.Spec.Template.Spec.Tolerations = tolerations
+	}
+
+	return ds
+}
+
+// mergeMaps adds all key/value pairs from the provided map to the current map
+// if those keys do not already exist in the current map.
+func mergeMaps(current map[string]string, provided map[string]string) {
+	for k, v := range provided {
+		// Only copy this key/value pair from the provided map if it does
+		// not already exist.
+		if _, ok := current[k]; !ok {
+			current[k] = v
+		}
+	}
+}
+
+// mergeContainers copies the ResourceRequirements from the provided containers
+// to the current corev1.Containers.
+func mergeContainers(current []corev1.Container, provided []container) {
+	for _, pc := range provided {
+		// Skip if no resources defined.
+		if pc.Resources == nil {
+			continue
+		}
+
+		var index int
+		var found bool
+		for i, cc := range current {
+			if cc.Name == pc.Name {
+				index = i
+				found = true
+				break
+			}
+		}
+		if found {
+			current[index].Resources = *pc.Resources
+		}
```

Comment:
Because this is called from the node render, there isn't a good way to bubble up errors. The render.Component interface: https://github.com/tigera/operator/blob/master/pkg/render/component.go#L32

We'll have CRD validation and installation validation from the core controller at least.

Reply:
> should only happen in the case of a code bug, but would be good to fail loudly rather than silently.

Sorry, what I meant by this is that it should never happen assuming there are no bugs in the calling code - we don't need to bubble an error up, but we should at least log an error, e.g., `log.Error("BUG: hit unexpected line")`